src/awkward/_nplikes/virtual.py

```diff
@@ -0,0 +1,290 @@
+    Some operations (such as trivial slicing) maintain virtualness and return a new virtual array.
+    Others are required to access the underlying data of the array and therefore materialize it.
+    The materialized arrays are cached on themselves in the `_array` property.
+    Subsequent virtual arrays that originate from some virtual array will hit the cache of their parents if there is any.
+    """
+
+    def __init__(
+        self,
+        nplike: NumpyLike,
+        shape: tuple[ShapeItem, ...],
+        dtype: DTypeLike,
+        generator: Callable[[], ArrayLike],
+    ) -> None:
+        if not isinstance(nplike, (ak._nplikes.numpy.Numpy, ak._nplikes.cupy.Cupy)):
+            raise TypeError(
+                f"Only numpy and cupy nplikes are supported for {type(self).__name__}. Received {type(nplike)}"
+            )
+        if any(not is_integer(dim) for dim in shape):
+            raise TypeError(
+                f"Only shapes of integer dimensions are supported for {type(self).__name__}. Received shape {shape}."
+            )
+
+        # array metadata
+        self._nplike = nplike
+        self._shape = shape
+        self._dtype = np.dtype(dtype)
+        self._array: Sentinel | ArrayLike = UNMATERIALIZED
+        self._generator = generator
+
+    @property
+    def dtype(self) -> DType:
+        return self._dtype
+
+    @property
+    def shape(self) -> tuple[ShapeItem, ...]:
+        return self._shape
+
+    @property
+    def ndim(self) -> int:
+        return len(self._shape)
+
+    @property
+    def size(self) -> ShapeItem:
+        if len(self._shape) == 0:
+            return 1
+        return reduce(mul, self._shape)
+
+    @property
+    def nbytes(self) -> ShapeItem:
+        if self.is_materialized:
+            return self._array.nbytes  # type: ignore[union-attr]
+        return 0
+
+    @property
+    def strides(self) -> tuple[ShapeItem, ...]:
+        return self.materialize().strides  # type: ignore[attr-defined]
+
+    def materialize(self) -> ArrayLike:
+        if self._array is UNMATERIALIZED:
+            array = self._nplike.asarray(self.generator())
+            if self._shape != array.shape:
+                raise TypeError(
+                    f"{type(self).__name__} had shape {self._shape} before materialization while the materialized array has shape {array.shape}"
+                )
+            if self._dtype != array.dtype:
+                raise TypeError(
+                    f"{type(self).__name__} had dtype {self._dtype} before materialization while the materialized array has dtype {array.dtype}"
+                )
+            self._array = array
+        return self._array  # type: ignore[return-value]
+
+    def dematerialize(self) -> None:
+        self._array = UNMATERIALIZED
+
+    @property
+    def is_materialized(self) -> bool:
+        return self._array is not UNMATERIALIZED
+
+    @property
+    def T(self):
+        if self.is_materialized:
```

Comment:
For these internal methods, I'd prefer to reduce the abstraction and directly test
```suggestion
        if self._array is not UNMATERIALIZED:
```